FIX: Ensure CORS rules exist for S3 using rake task

martin-brennan opened the pull request:
This commit introduces a new s3:ensure_cors_rules rake task
that is run as a prerequisite to s3:upload_assets. This rake
task calls out to the S3CorsRulesets class to ensure that
the 3 relevant sets of CORS rules are applied, depending on
site settings:

* assets
* direct S3 backups
* direct S3 uploads

This works for both Global S3 settings and Database S3 settings
(the latter set directly via SiteSetting).

As it is, only one rule can be applied, which is generally
the assets rule as it is called first. This commit changes
the ensure_cors! method to be able to apply new rules as
well as the existing ones.

This commit also slightly changes the existing rules to cover
direct S3 uploads via uppy, especially multipart, which requires
some more headers.

**NOTE: Please do not merge this, I will merge it once some internal infrastructure has shifted.**

tgxworld commented on lib/s3_cors_rulesets.rb:
```diff
@@ -10,8 +12,55 @@ class S3CorsRulesets
     allowed_headers: ["*"],
-    allowed_methods: ["PUT"],
+    expose_headers: ["ETag"],
+    allowed_methods: ["GET", "HEAD", "PUT"],
```
Would like to clarify why we need `GET` and `HEAD` as compared to just `PUT` previously.

martin-brennan replied:
Will discuss this internally as well, but it is because this will be needed for direct S3 backup multipart uploads, and I would prefer to apply all these new rules at the same time.

tgxworld commented on lib/s3_cors_rulesets.rb:
```diff
@@ -10,8 +12,55 @@ class S3CorsRulesets
+  # The use_db_s3_config param comes from ENV, and if the S3 client
+  # is not provided it is initialized by the S3Helper.
+  def self.sync(use_db_s3_config:, s3_client: nil)
+    return unless SiteSetting.s3_install_cors_rule && \
+      (GlobalSetting.use_s3? || SiteSetting.enable_s3_uploads)
```
```suggestion
    return if !SiteSetting.s3_install_cors_rule
    return if ! (GlobalSetting.use_s3? || SiteSetting.enable_s3_uploads)
```

Maybe I'm getting older but I feel like this is much easier to understand.

martin-brennan replied:
No you're right, not sure why I committed the sin of using `unless` XD

tgxworld commented on lib/s3_cors_rulesets.rb:
```diff
@@ -10,8 +12,55 @@ class S3CorsRulesets
+    )
+    assets_rules_applied = s3_helper.ensure_cors!([S3CorsRulesets::ASSETS])
+
+    if SiteSetting.enable_backups? && SiteSetting.backup_location == BackupLocationSiteSetting::S3
```
Hmm does this mean that we'll need to redeploy when `SiteSetting.backup_location` is changed to S3 before the CORS rules are applied?

tgxworld commented on lib/s3_cors_rulesets.rb:
```diff
@@ -10,8 +12,55 @@ class S3CorsRulesets
+      backup_rules_applied = backup_s3_helper.ensure_cors!([S3CorsRulesets::BACKUP_DIRECT_UPLOAD])
+    end
+
+    if SiteSetting.enable_direct_s3_uploads
```
Same concern as above where I think we might need to redeploy before the CORS rules are applied when this site setting is changed.

tgxworld commented on lib/s3_helper.rb:
```diff
@@ -40,6 +40,24 @@ def initialize(s3_bucket_name, tombstone_prefix = '', options = {})
       end
   end
 
+  def self.build_from_config(use_db_s3_config: false, for_backup: false, s3_client: nil)
+    bucket, options =
+      if use_db_s3_config
+        [
+          for_backup ? SiteSetting.s3_backup_bucket.downcase : SiteSetting.s3_upload_bucket.downcase,
+          S3Helper.s3_options(SiteSetting)
+        ]
+      else
+        [
+          for_backup ? GlobalSetting.s3_backup_bucket.downcase : GlobalSetting.s3_bucket.downcase,
+          S3Helper.s3_options(GlobalSetting)
+        ]
+      end
+    options[:client] = s3_client if s3_client.present?
+
+    S3Helper.new(bucket, '', options)
+  end
```
```suggestion
  def self.build_from_config(use_db_s3_config: false, for_backup: false, s3_client: nil)
    klass = use_db_s3_config ? SiteSetting : GlobalSetting
    
    bucket, options = [
      for_backup ? klass.s3_backup_bucket.downcase : klass.s3_upload_bucket.downcase,
      S3Helper.s3_options(klass)
    ]
    
    options[:client] = s3_client if s3_client.present?

    S3Helper.new(bucket, '', options)
  end
```

martin-brennan replied:
Unfortunately this does not _quite_ work because the S3 bucket property is slightly different between global + site:

![image](https://user-images.githubusercontent.com/920448/140009103-d040cba5-c1cf-40dd-a550-bd6bb1c2d8b5.png)

martin-brennan replied:
Will try to clean it up a bit though.

tgxworld commented on spec/lib/s3_cors_rulesets_spec.rb:
```diff
@@ -0,0 +1,209 @@
+        global_setting :s3_region, "us-west-2"
+      end
+
+      it "does nothing if !s3_install_cors_rule" do
+        SiteSetting.s3_install_cors_rule = false
+        sync_rules
+      end
```
Minor but perhaps we can assert for the results returned from `sync_rules` here? Otherwise, it seems strange that there are no assertion in this test.

martin-brennan replied:
Yes fair enough

tgxworld commented on spec/lib/s3_cors_rulesets_spec.rb:
```diff
@@ -0,0 +1,209 @@
+        )
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-backup-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::BACKUP_DIRECT_UPLOAD
+            ]
+          }
+        )
+        result = sync_rules
+        expect(result[:assets_rules_applied]).to eq(true)
+        expect(result[:direct_upload_rules_applied]).to eq(false)
+        expect(result[:backup_rules_applied]).to eq(true)
+      end
+
+      it "applies the ASSETS rules and the DIRECT_UPLOAD rules when S3 direct uploads are enabled" do
+        SiteSetting.enable_direct_s3_uploads = true
+
+        client.stub_responses(:get_bucket_cors, {})
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::ASSETS
+            ]
+          }
+        )
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::DIRECT_UPLOAD
+            ]
+          }
+        )
+        result = sync_rules
+        expect(result[:assets_rules_applied]).to eq(true)
+        expect(result[:direct_upload_rules_applied]).to eq(true)
+        expect(result[:backup_rules_applied]).to eq(false)
+      end
```
I kind of feel like all these tests are not necessary since they are eventually covered by the "applies all rules if everything is set up" test below.

martin-brennan replied:
I would prefer to get rid of the "applies all rules" test then, I want to have explicit tests for each of the "configured" rules cause they have distinct logic.

tgxworld commented on spec/lib/s3_cors_rulesets_spec.rb:
```diff
@@ -0,0 +1,209 @@
+        expect(result[:backup_rules_applied]).to eq(false)
+      end
+
+      def setup_backups
+        SiteSetting.enable_backups = true
+        SiteSetting.s3_backup_bucket = "s3-backup-upload-bucket"
+        SiteSetting.s3_use_iam_profile = true
+        SiteSetting.backup_location = BackupLocationSiteSetting::S3
+      end
```
Hmm do we need this when the global settings have already been configured?

martin-brennan replied:
Yes we do, see:

https://github.com/discourse/discourse/blob/fada309c9a5c70c2c757a09b001ecfd9b568422f/lib/site_settings/validations.rb#L159-L167

I agree it is a little strange, confused me a bit why that was needed but the backup bucket stuff is not loaded into GlobalSetting

martin-brennan commented:
@tgxworld just addressing these concerns:

> Hmm does this mean that we'll need to redeploy when SiteSetting.backup_location is changed to S3 before the CORS rules are applied?

> Same concern as above where I think we might need to redeploy before the CORS rules are applied when this site setting is changed.

I don't think this will be a problem on our hosting. I think we would rarely if ever change the backup location (we use S3 for all hosting) and the direct S3 uploads site setting is currently somewhat experimental; once it is more stable I can see us enabling that by default on our hosting as well.

If either of them do change, the operator can just run the `s3:ensure_cors` rake task to even the rules out.

tgxworld commented:
> If either of them do change, the operator can just run the `s3:ensure_cors` rake task to even the rules out.

I think my concern here is that the syncing here is based on site settings which are usually done via the admin interface but for it to work we need to run a rake task manually. For people managing their own setups, I think we either need to indicate that in the site setting description and change the CORS rule when the site setting changes.

martin-brennan commented:
> For people managing their own setups, I think we either need to indicate that in the site setting description and change the CORS rule when the site setting changes.

Yes agreed we need to do something slightly different for self hosters, that can be a future PR though. To counter this for self hosters, I will just revert this change to not call `ensure_cors!` inside the S3BackupStore:

![image](https://user-images.githubusercontent.com/920448/140013063-be6e818f-ee7b-4da3-b9b5-3950b5d6b23a.png)

Later on we can take that out in favour of the auto-change on site setting change.
